Blog post for build scheduler features (#170)
shipwright-io/website: content/docs/blog/posts/2025-5-20-build-scheduler-features.md

---
title: "Shipwright build scheduler features"
date: 2025-05-20T11:39:00-06:00

> draft: true

Reviewer: Change this to `false` so it renders in the deploy preview:

```suggestion
draft: false
```

author: "Dylan Orzel ([@dorzel](https://github.com/dorzel))"
---

## New Build Scheduler Features

> A new set of build scheduling features introduced in [v0.15](https://shipwright.io/docs/blog/posts/2025-02-28-release-v0.15) have recently been merged into Shipwright that allows users to specify node selectors, custom schedulers, and tolerations for builds.

Reviewer: nit to clarify that this isn't exactly recent:

```suggestion
A new set of build scheduling features introduced in [v0.15](https://shipwright.io/docs/blog/posts/2025-02-28-release-v0.15) allows users to specify node selectors, custom schedulers, and tolerations for builds.
```

> These make it easier to schedule builds when using Shipwright on clusters with nodes of multiple architectures, use a scheduler that is tuned to a certain workflow, or just more general control of which nodes builds get scheduled on.

Reviewer: nit - we can assume "when using Shipwright", and fix repeated words:

```suggestion
These make it easier to schedule builds on clusters with nodes of multiple CPU architectures, use a scheduler that is tuned to a certain workflow, or just more general control of which nodes builds run on.
```

and see that the created BuildRun now has a nodeSelector specified and has been scheduled on the arm64 node:

```yaml
apiVersion: shipwright.io/v1beta1
kind: BuildRun
metadata:
  creationTimestamp: "2025-05-20T17:27:01Z"
  generateName: test-golang-build-
  generation: 1
  labels:
    build.shipwright.io/generation: "1"
    build.shipwright.io/name: test-golang-build
  name: test-golang-build-glghq
  namespace: default
  resourceVersion: "31512392"
  uid: 19f02c99-1d24-44c0-a5f9-ab544fdf55ae
spec:
  build:
    name: test-golang-build
  nodeSelector:
    kubernetes.io/arch: arm64
```

```bash
$ kubectl get pod test-golang-build-glghq-tfhnn-pod -o jsonpath='{.spec.nodeName} {.spec.nodeSelector}'

arm-node.compute.internal {"kubernetes.io/arch":"arm64"}

$ kubectl get node arm-node.compute.internal -o jsonpath='{.metadata.labels.kubernetes\.io/arch}'

arm64
```

### Example with `shp build create` and `--scheduler-name`

This time we'll specify a scheduler name `test-scheduler` that is assumed to be a custom scheduler that has been deployed to the cluster already.

Instead of specifying these options when running the build, we can also specify them when creating the build:

```bash

> $ shp build create test-golang-build --output-image=kind.local/test/test-golang-build --source-git-url=https://github.com/shipwright-io/sample-go --source-context-dir=docker-build --strategy-name=buildah-shipwright-managed-push --scheduler-name=test-scheduler

Reviewer: nit - use newlines since we have so many flags:

```suggestion
$ shp build create test-golang-build \
  --output-image=kind.local/test/test-golang-build \
  --source-git-url=https://github.com/shipwright-io/sample-go \
  --source-context-dir=docker-build \
  --strategy-name=buildah-shipwright-managed-push \
  --scheduler-name=test-scheduler
```

```

and see that the scheduler name appears in the Build yaml:

```yaml
apiVersion: shipwright.io/v1beta1
kind: Build
metadata:
  creationTimestamp: "2025-05-20T17:49:07Z"
  generation: 1
  name: test-golang-build
  namespace: default
  resourceVersion: "31518385"
  uid: bd2333fb-71eb-4228-bc5f-5a466032fcc5
spec:
  output:
    image: kind.local/test/test-golang-build
  schedulerName: test-scheduler
  source:
    contextDir: docker-build
    git:
      url: https://github.com/shipwright-io/sample-go
    type: Git
  strategy:
    kind: ClusterBuildStrategy
    name: buildah-shipwright-managed-push
status:
  message: all validations succeeded
  reason: Succeeded
  registered: "True"
```

After running the build with `shp build run test-golang-build`, we see that the schedulerName got picked up by the build pod:

```bash
$ kubectl get pods test-golang-build-j9vbd-qr9cr-pod -o jsonpath='{.spec.schedulerName}'

test-scheduler
```

### Example with setting tolerations on a `BuildRun`

We'll start with the same example build as above:

```bash

> $ shp build create test-golang-build --output-image=kind.local/test/test-golang-build --source-git-url=https://github.com/shipwright-io/sample-go --source-context-dir=docker-build --strategy-name=buildah-shipwright-managed-push

Reviewer: nit - ditto on newlines:

```suggestion
$ shp build create test-golang-build \
  --output-image=kind.local/test/test-golang-build \
  --source-git-url=https://github.com/shipwright-io/sample-go \
  --source-context-dir=docker-build \
  --strategy-name=buildah-shipwright-managed-push
```
